### ISSUE TITLE
Work is added to migrations for model/database properties that are identical (so should not be added)

### ISSUE THREAD

**Issue by mnmr** [aspnet/EntityFrameworkCore#11297]
When I add a new migration, EF Core (latest stable 2.1.0 preview/CI build) keeps adding migrations that do not change either model or database. Even if I apply the suggested migrations, next time I add a migration they're being added again.

The following shows the original init definition for the type:
```C#
migrationBuilder.CreateTable(
            name: "UserTradingSettings",
                columns: table => new
                {
                    Id = table.Column<long>(nullable: false)
                        .Annotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn),
                    Name = table.Column<string>(maxLength: 50, nullable: false),
                    Sport = table.Column<int>(nullable: true),
                    UserId = table.Column<long>(nullable: false),
                    Value = table.Column<string>(maxLength: 2000, nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_UserTradingSettings", x => x.Id);
                    table.ForeignKey(
                        name: "FK_dbo.UserTradingSettings_dbo.Users_UserId",
                        column: x => x.UserId,
                        principalTable: "Users",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Cascade);
                });
```

This is the current snapshot for the type:
```C#
modelBuilder.Entity("BP.Domain.Contracts.Entities.UserTradingSetting", b =>
                {
                    b.Property<long>("Id")
                        .ValueGeneratedOnAdd();

                    b.Property<string>("Name")
                        .IsRequired()
                        .HasMaxLength(50)
                        .IsUnicode(true);

                    b.Property<int>("Sport");

                    b.Property<long>("UserId");

                    b.Property<string>("Value")
                        .IsRequired()
                        .HasMaxLength(2000)
                        .IsUnicode(true);

                    b.HasKey("Id");

                    b.HasIndex("UserId")
                        .HasName("IX_UserId");

                    b.ToTable("UserTradingSettings");
                });
```

This is what the current table definition looks like:
![image](https://user-images.githubusercontent.com/1616137/37520408-281f73b0-291d-11e8-8ff5-4163bdfa815a.png)

And this is the erroneous change that EF adds to new migrations:
```C#
migrationBuilder.AlterColumn<int>(
                name: "Sport",
                table: "UserTradingSettings",
                nullable: true,
                oldClrType: typeof(int));
```

The class itself looks like this (ESport is an enumeration):
```C#
    public class UserTradingSetting : IEntity
    {
        public long Id { get; set; }
        public ESport? Sport { get; set; }
        public string Name { get; set; }
        public string Value { get; set; }
        public long UserId { get; set; }

        public User User { get; set; }        
    }
```

And the mapping for the class:
```
    modelBuilder.HasKey(x => x.Id);
    modelBuilder.Property(e => e.Id).IsRequired();
    modelBuilder.Property(e => e.Sport);
    modelBuilder.Property(e => e.UserId).IsRequired();

    modelBuilder.Property(e => e.Name).IsRequired().IsUnicode(true).HasMaxLength(50);
    modelBuilder.Property(e => e.Value).IsRequired().IsUnicode(true).HasMaxLength(2000);

    modelBuilder.HasOne(e => e.User).WithMany().HasForeignKey(e => e.UserId).IsRequired()
        .OnDelete(DeleteBehavior.Cascade).HasConstraintName("FK_dbo.UserTradingSettings_dbo.Users_UserId");

    modelBuilder.HasIndex(e => e.UserId).HasName("IX_UserId");
```

### Further technical details
EF Core version: 2.1.0-preview2-30355
Database Provider: Microsoft.EntityFrameworkCore.SqlServer

**Comment by bricelam** [comment]
The snapshot is missing `Nullable` after converting to the enum's underlying type:
``` csharp
b.Property<int>("Sport");
```

**Comment by mnmr** [comment]
There is no method (or extension) that I could find called `Nullable`.

As far as I understand, the omission of .IsRequired() on the property should instruct EF to make the column nullable. As such, the mapping already specifies that the property should be nullable.

**Comment by bricelam** [comment]
The snapshot should be...

``` csharp
b.Property<int?>("Sport");
```

**Comment by mnmr** [comment]
Right. However, I am not the author of the snapshot class - it is auto-generated. If I modify the snapshot (to fix the mappings) and create a new (empty) migration, the generated snapshot is once again back to being incorrect and I'd need to reapply the fix.

It's therefore not something I should fix in my code, but rather something that needs to be fixed in the code that emits the snapshot in the first place.

**Comment by bricelam** [comment]
Yes. This is a bug. We'll fix it before the final 2.1.0 release.

There's no good workaround except to remove the `AlterColumn` call every time you generate a new migration.

**Comment by mnmr** [comment]
Great, thanks!
